lib/core/src/display/navigation/event_handler.rs

```diff
@@ -0,0 +1,326 @@
+        };
+        add_mouse_event(&target  , "mouseup"   , closure.clone())?.forget();
+        add_mouse_event(&document, "mouseleave", closure        )?.forget();
+        Ok(())
+    }
+
+    /// Initialize wheel event callback.
+    fn initialize_wheel_event
+    (self:&Rc<Self>, target:&EventTarget) -> Result<()>{
+        let ehandler = self.clone();
+        let closure = move |event:WheelEvent| {
+            let ime     = &mut ehandler.data.borrow_mut();
+            let ime     = ime as &mut InternalMouseEvent;
+            let delta_y = event.delta_y() as f32;
+            match ime {
+                InternalMouseEvent::Start(mouse_start) => {
+                    let position    = mouse_start.position;
+                    let mouse_wheel = MouseWheel { position, delta_y };
+                    *ime            = InternalMouseEvent::Wheel(mouse_wheel);
+                },
+                InternalMouseEvent::Move(mouse_move) => {
+                    let position    = mouse_move.current;
+                    let mouse_wheel = MouseWheel { position, delta_y };
+                    *ime            = InternalMouseEvent::Wheel(mouse_wheel);
+                },
+                InternalMouseEvent::Wheel(mouse_wheel) => {
+                    mouse_wheel.delta_y = delta_y;
+                },
+                InternalMouseEvent::None => ()
+            }
+        };
+        add_wheel_event(&target, closure)?.forget();
+        Ok(())
+    }
+
+    /// Checks if EventHandler has an event.
+    pub fn poll(&self) -> Event {
+        let   event : &mut InternalMouseEvent = &mut self.data.borrow_mut();
+        match event.consume() {
+            InternalMouseEvent::Start(_)         => Event::Start,
+            InternalMouseEvent::Move(mouse_move) => {
+                if let Some(start) = &mouse_move.start {
+                    match start.button {
+                        MMB => {
+                            let mut movement   =  mouse_move.current;
+                                    movement  -=  mouse_move.previous;
+                                    movement.x = -movement.x;
+                            Event::Pan(PanEvent { movement })
+                        },
+                        RMB => {
+                            let focus  = start.position;
+                            let mut amount  = mouse_move.current.y;
+                                    amount -= mouse_move.previous.y;
+                            Event::Zoom(ZoomEvent { focus, amount })
+                        },
+                        _ => Event::None
+                    }
+                } else {
+                    Event::None
+                }
+            },
+            InternalMouseEvent::Wheel(mouse_wheel) => {
+                let focus  = mouse_wheel.position;
+                let amount = mouse_wheel.delta_y;
+                Event::Zoom(ZoomEvent { focus, amount })
+            }
+            _ => Event::None
+        }
+    }
+}
```

Comment:
The whole code lacks proper documentation. I spend way too much time to understand what really is the `EventHandler` structure and how it works. The docs for `EventHandler` should be really descriptive and tell how it works and why. However, I did the job so now I understand that:

- You connect closures to DOM events and in case a dom event is emitted, you transform the event to some "internal" event and you store the last event in the `EventHandler` structure.
- Sometimes (like every frame) you use `poll` to check what was the last event in `EventHandler` to use it.

If that's correct, then this design is bad and needs to be rewritten, as it can accidentally drop events. You should never store the LAST event in `EventHandler` because in a single frame several events may be emitted and you will drop them. This is very bad.

Reply:
This will be addressed in the refactoring.

Reply:
Wait, what? What happens if I press LMB and then RMB in Chrome so fast they are in single frame? No way they will be dropped.

Reply:
`poll` now pops from an event queue and we are not losing any events.